server/region_handler.go

```diff
@@ -0,0 +1,512 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"math"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/juju/errors"
+	"github.com/pingcap/pd/pd-client"
+	"github.com/pingcap/tidb"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/domain"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/store/tikv"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/util/codec"
+	goctx "golang.org/x/net/context"
+)
+
+// RegionsHTTPRequest is for regions http request.
+type RegionsHTTPRequest struct {
+	resp        http.ResponseWriter
+	req         *http.Request
+	server      *Server
+	params      map[string]string
+	reqID       string
+	regionCache *tikv.RegionCache
+	bo          *tikv.Backoffer
+	domain      *domain.Domain
+}
+
+// NewRegionsHTTPRequest create a RegionsHTTPRequest
+func NewRegionsHTTPRequest(s *Server, w http.ResponseWriter, req *http.Request,
+	pdClient *pd.Client) (regionReq *RegionsHTTPRequest, err error) {
+	regionReq = &RegionsHTTPRequest{
+		server: s,
+		req:    req,
+		resp:   w,
+		params: mux.Vars(req),
+		//TODO genuuid
+		reqID: fmt.Sprintf("http_%d", time.Now().UnixNano()),
+	}
+	err = regionReq.init(pdClient)
+	return regionReq, errors.Trace(err)
+}
+
+func (req *RegionsHTTPRequest) writeResponse(data interface{}, err error) {
+	var retItem *RegionsResponse
+	if err != nil {
+		retItem = NewRegionsResponseWithError(req.reqID, err)
```

Comment:
For client, we can check return status directly. 

Here some error you return json, but some you return the error string, not unified. So now if we don't need it, remove it, if we need it in the future, add it.